### ISSUE TITLE
Add a priority to a backendRef

### ISSUE THREAD

**Issue by arkodg** [kubernetes-sigs/gateway-api#1802]
**What would you like to be added**:
Add a field within [backendRef](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.BackendRef) allowing the user to specify the priority of a backend. A algo for traffic distribution is the following
1. Pick all the healthy backends with the highest priority level
2. Distribute traffic across them based on the `weight` ratio
3. If no healthy backend exists for the highest priority, perform above steps for the next priority

**Why this is needed**:
As a user I would like to specify a default / fallback / back up backend in case my main backend is no longer active
also relates to https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#default-backend

**Comment by shaneutt** [comment]
To me this is reminiscent of https://github.com/kubernetes/kubernetes/issues/56440 (also relatedly, https://github.com/kubernetes/kubernetes/issues/115917) where we were discussing fallback positions for traffic when things go wrong and we were starting to discuss whether Gateway API might be a good place for such a thing (as opposed to `Service` which we're trying to avoid adding anything to). Was that intentional? If not give those a read and let me know what you think :thinking:

**Comment by arkodg** [comment]
thanks for sharing the links @shaneutt, the intent is similar (sending traffic to a backup backend), but that intent needs to be defined at the `backendRef` level rather than the specific type of backend (`Service`) so traffic can be prioritized to different backend `Kind`s

**Comment by howardjohn** [comment]
Similar use case is things like "give priority to local endpoints _within_ a service" (where local could be node, region, zone, cluster, mesh, network, or arbitrary things). Perhaps that is a different case though

**Comment by howardjohn** [comment]
Another alternative is to make a non-Service backend that represents some "Composite" backend. Which could be implementation specific or not.  Related: https://github.com/kubernetes-sigs/gateway-api/discussions/1764

**Comment by shaneutt** [comment]
Seems like between us and the fact that the `kubernetes/kubernetes` has a relevant similar use case which @arkodg identifies with, we would just need someone to champion this issue, bring it up at the community meetings, and start the GEP. Are either of you particularly interested in leading on this one?

/help

**Comment by arkodg** [comment]
happy to pick this one up @shaneutt and raise a GEP

**Comment by shaneutt** [comment]
Yes, but my only hesitation would be that it doesn't seem required for GA, so as per the new GEP docs you can definitely start a discussion, but you may want to add language to the discussion pointing out that it's understood this might not get much attention until GA, and it may be a couple months before it really starts moving.

/assign @arkodg

**Comment by arkodg** [comment]
being discussed in https://github.com/kubernetes-sigs/gateway-api/discussions/2304

**Comment by arkodg** [comment]
thinking out loud, this feature will also need an accompanying 'healthCheck' field to decide how to compute the health of a backend

**Comment by mikemorris** [comment]
@kflynn @howardjohn is this worth revisiting as we consider planning for future "GAMMA" work? It seems like several Gateway API implementations have some sort of equivalent bespoke functionality to this proposed feature.

**Comment by howardjohn** [comment]
I don't think there is anything particularly connecting this to mesh; many organizations actually only do failover at ingress. But could be worth visiting, though I suspect this may be one of those areas where there are a lot of implementation-specific differentiators.

There is also some first-class Kubernetes support in `trafficDistribution`

**Comment by keithmattix** [comment]
> Another alternative is to make a non-Service backend that represents some "Composite" backend. Which could be implementation specific or not.  Related: https://github.com/kubernetes-sigs/gateway-api/discussions/1764

/cc @grosskur @robscott re the early discussions we've had oj this kind of resource.
